sophiread/SophireadCLI/src/sophiread_core.cpp

```diff
@@ -0,0 +1,382 @@
+  auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
+  spdlog::info("Read raw data: {} s", elapsed / 1e6);
+
+  return raw_data;
+}
+
+/**
+ * @brief Timed find TPX3H.
+ *
+ * @param[in] rawdata
+ * @return std::vector<TPX3>
+ */
+std::vector<TPX3> timedFindTPX3H(const std::vector<char> &rawdata) {
+  auto start = std::chrono::high_resolution_clock::now();
+  auto batches = findTPX3H(rawdata);
+  auto end = std::chrono::high_resolution_clock::now();
+  auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
+  spdlog::info("Locate all headers: {} s", elapsed / 1e6);
+
+  return batches;
+}
+
+/**
+ * @brief Timed locate timestamp.
+ *
+ * @param[in, out] batches
+ * @param[in] rawdata
+ */
+void timedLocateTimeStamp(std::vector<TPX3> &batches, const std::vector<char> &rawdata) {
+  auto start = std::chrono::high_resolution_clock::now();
+  unsigned long tdc_timestamp = 0;
+  unsigned long long gdc_timestamp = 0;
+  unsigned long timer_lsb32 = 0;
+  for (auto &tpx3 : batches) {
+    updateTimestamp(tpx3, rawdata, tdc_timestamp, gdc_timestamp, timer_lsb32);
+  }
+  auto end = std::chrono::high_resolution_clock::now();
+  auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
+  spdlog::info("Locate all timestamps: {} s", elapsed / 1e6);
+}
+
+/**
+ * @brief Timed hits extraction and clustering via multi-threading.
+ *
+ * @param[in, out] batches
+ * @param[in] rawdata
+ * @param[in] config
+ */
+void timedProcessing(std::vector<TPX3> &batches, const std::vector<char> &raw_data, const IConfig &config) {
+  auto start = std::chrono::high_resolution_clock::now();
+  tbb::parallel_for(tbb::blocked_range<size_t>(0, batches.size()), [&](const tbb::blocked_range<size_t> &r) {
+    // Define ABS algorithm with user-defined parameters for each thread
+    auto abs_alg_mt =
+        std::make_unique<ABS>(config.getABSRadius(), config.getABSMinClusterSize(), config.getABSSpiderTimeRange());
+
+    for (size_t i = r.begin(); i != r.end(); ++i) {
+      auto &tpx3 = batches[i];
+      extractHits(tpx3, raw_data);
+
+      abs_alg_mt->reset();
+      abs_alg_mt->set_method("centroid");
+      abs_alg_mt->fit(tpx3.hits);
+
+      tpx3.neutrons = abs_alg_mt->get_events(tpx3.hits);
+    }
+  });
+  auto end = std::chrono::high_resolution_clock::now();
+  auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
+  spdlog::info("Process all hits -> neutrons: {} s", elapsed / 1e6);
+}
+
+/**
+ * @brief Timed save hits to HDF5.
+ *
+ * @param[in] out_hits
+ * @param[in] hits
+ */
+void timedSaveHitsToHDF5(const std::string &out_hits, std::vector<TPX3> &batches) {
+  auto start = std::chrono::high_resolution_clock::now();
+  // move all hits into a single vector
+  std::vector<Hit> hits;
+  for (const auto &tpx3 : batches) {
+    auto tpx3_hits = tpx3.hits;
+    hits.insert(hits.end(), tpx3_hits.begin(), tpx3_hits.end());
```

Comment:
See [`std::copy`](https://en.cppreference.com/w/cpp/algorithm/copy) and use `.cbegin()` and `.cend()`.
```cpp
const auto tpx3_hits = tpx3.hits; // a reference if the compiler will let you
hits.copy(tpx3_hits.cbegin(), tpx3_hits.cend(), std::back_inserter(hits));
```
Otherwise, I'm not sure what std::incert

Reply:
switch to use `const auto` and `&` to avoid local copy